node/rpc/src/gas_api.rs

```diff
@@ -0,0 +1,189 @@
+        .ok_or("cant get heaviest tipset")?;
+
+    for _ in 0..(nblocksincl * 2) {
+        if ts.parents().cids().is_empty() {
+            break;
+        }
+        let pts = chain::tipset_from_keys(data.state_manager.get_block_store_ref(), ts.parents())?;
+        blocks += pts.blocks().len();
+        let msgs = chain::messages_for_tipset(data.state_manager.get_block_store_ref(), &pts)?;
+
+        prices.append(
+            &mut msgs
+                .iter()
+                .map(|msg| GasMeta {
+                    price: msg.gas_premium().clone(),
+                    limit: msg.gas_limit(),
+                })
+                .collect(),
+        );
+        ts = pts;
+    }
+
+    prices.sort_by(|a, b| b.price.cmp(&a.price));
+    // TODO: From lotus, account for how full blocks are
+    let mut at = BLOCK_GAS_TARGET * blocks as i64 / 2;
+    let mut prev = BigInt::zero();
+    let mut premium = BigInt::zero();
+
+    for price in prices {
+        at -= price.limit;
+        if at > 0 {
+            prev = price.price;
+            continue;
+        }
+        if prev == 0.into() {
+            let ret: BigInt = price.price + 1;
+            return Ok(ret.to_string());
+        }
+        premium = (&price.price + &prev) / 2 + 1
+    }
+
+    if premium == 0.into() {
+        premium = BigInt::from_f64(match nblocksincl {
+            1 => MIN_GAS_PREMIUM * 2.0,
+            2 => MIN_GAS_PREMIUM * 1.5,
+            _ => MIN_GAS_PREMIUM,
+        })
+        .ok_or("failed to convert gas premium f64 to bigint")?;
+    }
+
+    let precision = 32;
+
+    // mean 1, stddev 0.005 => 95% within +-1%
+    let noise: f64 = Normal::new(1.0, 0.005)
+        .unwrap()
+        .sample(&mut rand::thread_rng());
+    premium *= BigInt::from_f64(noise * (1 << precision) as f64)
+        .ok_or("failed to converrt gas premium f64 to bigint")?;
+    premium /= 1 << precision;
+
+    Ok(premium.to_string())
+}
+
+/// Estimate the gas limit
+pub(crate) async fn gas_estimate_gas_limit<DB, KS>(
+    data: Data<RpcState<DB, KS>>,
+    Params(params): Params<(UnsignedMessageJson, TipsetKeys)>,
+) -> Result<i64, JsonRpcError>
+where
+    DB: BlockStore + Send + Sync + 'static,
+    KS: KeyStore + Send + Sync + 'static,
+{
+    let (UnsignedMessageJson(mut msg), _) = params;
+    msg.set_gas_limit(BLOCK_GAS_LIMIT);
+    msg.set_gas_fee_cap((MINIMUM_BASE_FEE + 1).into());
+    msg.set_gas_premium(1.into());
+
+    let curr_ts = chain::get_heaviest_tipset(data.state_manager.get_block_store_ref())?
+        .ok_or("cant find the current heaviest tipset")?;
+    let from_a = data
+        .state_manager
+        .resolve_to_key_addr(msg.from(), &curr_ts)
+        .await?;
+
+    let pending = data.mpool.pending_for(&from_a).await;
+    let prior_messages: Vec<ChainMessage> = match pending {
+        Some(messages) => messages.into_iter().map(ChainMessage::Signed).collect(),
+        None => vec![],
+    };
```

Comment:
I don't know how you feel about this but it could also be written as 

pending.map(|s| s.into_iter().map(ChainMessage::Signed).collect()).unwrap_or_default()

Reply:
love this suggestion. thanks!